amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -212,24 +212,31 @@ private void updateLeasesForChildShards()
     private void garbageCollectLeaseForParentShard() throws DependencyException, ProvisionedThroughputException, InvalidStateException {
        final Lease parentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
 
-        if (parentShardLease.childShardIds().isEmpty()) {
+        if (CollectionUtils.isNullOrEmpty(parentShardLease.childShardIds())) {
             log.warn("No child shard state found in lease {} - not deleting lease.", parentShardLease);
         } else {
-            final Set<String> expectedLeaseKeys = parentShardLease.childShardIds().stream()
+            final Set<String> allChildShardLeases = parentShardLease.childShardIds().stream()
                     .map(c -> ShardInfo.getLeaseKey(shardInfo, c)).collect(Collectors.toSet());
-            final Set<String> leaseKeys = new HashSet<>();
+            final Set<String> processedChildShardLeases = new HashSet<>();
 
             for (String childShard : parentShardLease.childShardIds()) {
                 final String leaseKey = ShardInfo.getLeaseKey(shardInfo, childShard);
-                leaseKeys.add(leaseCoordinator.leaseRefresher().getLease(leaseKey).leaseKey());
+                final Lease lease = leaseCoordinator.leaseRefresher().getLease(leaseKey);
+
+                // We should only be deleting the parent shard lease if all of its children are currently being
+                // processed, i.e their checkpoint is not TRIM_HORIZON.
```

Comment:
Is this checkpoint the same as the checkpoint implemented by the customer? If so, what happens if they have a super delayed checkpointing strategy (e.g. checkpoint once a day or something), and so even if they are processing their checkpoint is still TRIM_HORIZON?

Other than a stale parent lease hanging around a little longer in DDB, are there any other side effects that could cause?

Reply:
Yeah this is the customer defined checkpointing, the worst case is that a parent shard will hang in DDB which will cause additional overhead in workers spinning up and realizing the shard is stale and spinning back down.